src/accountsdb/db.zig

```diff
@@ -1354,179 +1327,146 @@ pub const AccountsDB = struct {
-                                const info = in_file_account.account_info;
-                                var iterator = in_file_account.data.iterator();
-                                var hash = Hash.ZEROES;
-                                sig.core.account.hashAccount(
-                                    info.lamports,
-                                    &iterator,
-                                    &info.owner,
-                                    info.executable,
-                                    info.rent_epoch,
-                                    &in_file_account.store_info.pubkey,
-                                    &hash.data,
-                                );
-                                break :blk hash;
-                            },
-                            .unrooted_map => |unrooted_account| unrooted_account.hash(key),
-                        };
-                    }
-                }
+                // "mix in" account into hash
+                const latest_slot = slot: {
+                    // get the most recent state of the account
+                    const max_slot_ref = slotListMaxWithinBounds(
+                        ref_head.ref_ptr,
+                        maybe_min_slot,
+                        maybe_max_slot,
+                    ) orelse continue;
 
-                hashes.appendAssumeCapacity(account_hash);
-                local_total_lamports += lamports;
+                    const account = try self.getAccountFromRef(allocator, max_slot_ref);
+                    defer account.deinit(allocator);
+
+                    lt_hash.mixIn(account.ltHash(max_slot_ref.pubkey));
+
+                    total_lamports += account.lamports;
+
+                    if (max_slot_ref.slot == 0) continue;
+                    break :slot max_slot_ref.slot;
+                };
+
+                // "mix out" old version of account from hash
+                {
+                    // I expect this branch to not be taken for full snapshots, however it would
+                    // still be entirely legal.
+                    // The expected case is that we are validating an incremental snapshot, whose
+                    // lt_hash is initialised by the full snapshot's lt_hash. Accounts modified
+                    // in the incremental snapshot may also exist in the full snapshot, meaning we
+                    // are effectively removing the old entry.
+
+                    // remove previous state
+                    const previous_slot_ref = slotListMaxWithinBounds(
+                        ref_head.ref_ptr,
+                        null,
+                        latest_slot - 1,
+                    ) orelse continue;
+
+                    const account = try self.getAccountFromRef(allocator, previous_slot_ref);
+                    defer account.deinit(allocator);
+
+                    mix_out.mixIn(account.ltHash(previous_slot_ref.pubkey));
```

Comment:
Why not `lt_hash.mixOut`? the `mix_out` variable seems unnecessary.

Reply:
https://github.com/Syndica/sig/pull/993/commits/bf8250e6d704848a6cf5ffb04a743b6a782a809a

Think I followed agave here, removed it, we don't need it